src/templates/console.html

```diff
@@ -19,74 +19,79 @@
     <script>
       languagePluginLoader.then(() => {
         pyodide.runPython(`
-            from pyodide import console
+            import sys
+            import builtins
             import js
+            from reprlib import Repr
+            from pyodide.console import InteractiveConsole
 
 
-            class Console(console.InteractiveConsole):
+            # limiting output size to 1000 to avoid console hang
+            _repr = Repr()
+            _repr.maxdict = _repr.maxlist = _repr.maxtuple = _repr.maxset = _repr.maxfrozenset = _repr.maxdeque = _repr.maxarray = _repr.maxlong = _repr.maxstring = _repr.maxother = 10 ** 3
```

Comment:
I think 1000 elements might be too much, particularly for arrays represented as lists. It will still be very slow in the console.

Why not use values closer to [replib](https://docs.python.org/3/library/reprlib.html) defaults?

Reply:
Another option is to compute the entire repr and test the length of the output, then only take the beginning and end of the output, I've seen this done in some places. On my screen `console.html` is 34 lines high when it's full screen and at normal zoom. If we assume 34 * 80 characters wide, then we can fit 2720 characters per screen. Maybe we are targeting around 70% of that, so limit the output to 2000 characters? 

`reprlib` can both undershoot and overshoot this. For instance the following doesn't hang the console but it does generate 42000 characters out output, so ~20 screens worth of text:
```python
r = []
for i in range(5):
   r = [r,]*8
reprlib.repr(r)
```
On the other hand, by default `reprlib` only lists the first five elements of a 1d list, which is unacceptably stingy IMO (What if you want to see the first 10 elements to find a pattern? Sure you can explicitly ask for more to be printed, but I think this would be annoying.).

Also, the current console does not limit the width of the output region to 80 characters wide. On my screen in firefox with 100% zoom the console wraps at 140 characters.

Reply:
> Another option is to compute the entire repr and test the length of the output, then only take the beginning and end of the output

Maybe that would be easier indeed. As right if you have a list of a list of a dict, it's indeed hard to tell how long the final displayed output will be..

> Also, the current console does not limit the width of the output region to 80 characters wide. On my screen in firefox with 100% zoom the console wraps at 140 characters.

As far as I can tell the the behavior is identical to the classical Python REPL, in that the width limit is given by the window size? The only difference I see is that say when you print a very long list of str, Python REPL will not break strings in the middle, while this console will. But maybe we could look into it in a separate issue..

Reply:
>> does not limit width
>
> maybe we could look into it in a separate issue..

Agreed.

Reply:
Great, I'll do

```python
# limiting output size to REPR_LIMIT to avoid console hang
REPR_LIMIT = 1000
REPR_SPLIT = REPR_LIMIT // 4
def custom_repr(value):
    text = repr(value)
    if len(text) > REPR_LIMIT:
        text = f"{text[:REPR_SPLIT]}...{text[-REPR_SPLIT:]}"
    return text
```

This will avoid `repr` longer than 1000 and replace them by 503 character-long representation. Fill free to comment on the value I choose for `REPR_LIMIT` and `REPR_SPLIT`.

Reply:
Those look like reasonable values, though I think I'd set `REPR_SPLIT = REPR_LIMIT//2`? That way if it is 1010 characters it gets reduced to 1000 ish characters, it seems a little weird that if the output is 1000 characters the output is 1000 characters long but with 1001 characters it gets reduced to 500 ish characters.

It might also be good to do something like:

`text = f"{text[:REPR_SPLIT]}\n<<<<Long repl output truncated>>>>\n{text[-REPR_SPLIT:]}"` or something like that to make it a little to tell that the output has been truncated. Otherwise, the `...` will just appear in some random spot in the screen and probably not be all that noticeable. (If it's possible to make `<<<<Long repl output truncated>>>>` or whatever exactly it says a different color I think that'd be even better.)

Reply:
> it seems a little weird that if the output is 1000 characters the output is 1000 characters long but with 1001 characters it gets reduced to 500 ish characters.

That's what `numpy` do. I get full output for `np.array(range(1000))` and

```python
>>> np.array(range(1001))
array([   0,    1,    2, ...,  998,  999, 1000])
```

but I'm OK with what you say.

> It might also be good to do something like:

I'll try to add something like that.